commit 757a93810dd4251af3a336732dcfd0730a9996ef (refs/changes/12/812/10)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2017-05-31T13:12:27-07:00 (2 years, 4 months ago)
    
    HEAD-2325 Add Linux build support
    Reviewed by: Joshua M. Clulow <jmc@joyent.com>
    Reviewed by: Trent Mick <trentm@gmail.com>

diff --git a/Makefile b/Makefile
index c1e1fe2c..c1483da8 100644
--- a/Makefile
+++ b/Makefile
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright 2016 Joyent, Inc.
+# Copyright 2017 Joyent, Inc.
 #
 
 PERCENT := %
@@ -15,10 +15,10 @@ PERCENT := %
 # Files
 #
 
-ifeq ($(shell uname -s),Darwin)
-GREP = grep
-else
+ifeq ($(shell uname -s),SunOS)
 GREP = /usr/xpg4/bin/grep
+else
+GREP = grep
 endif
 
 BASH_FILES := \
diff --git a/bin/build-coal-image b/bin/build-coal-image
index 8ed5cbda..746ee871 100755
--- a/bin/build-coal-image
+++ b/bin/build-coal-image
@@ -105,8 +105,12 @@ if [[ -z $IMG_TMP_DIR ]]; then
 fi
 PLATFORM=$(uname -s)
 
-if [[ "$PLATFORM" != "Darwin" ]]; then
+if [[ "$PLATFORM" == "SunOS" ]]; then
     SUCMD='pfexec'
+elif [[ "$PLATFORM" == "Linux" ]]; then
+    SUCMD='sudo'
+    # Parted is required to find the partition offset in the image.
+    [[ $(which parted) ]] || fatal "Please install 'parted'"
 fi
 
 STAGE=${ROOT}/cache/stage_coal
@@ -183,13 +187,20 @@ function mount_image
             | awk '{ print $1 }'`
         MNT_DIR=`grep "Windows_FAT_32" /tmp/output.hdiattach.$$ \
             | awk '{ print $3 }'`
-
+    elif [[ "$PLATFORM" == "Linux" ]]; then
+        MNT_DIR="/tmp/sdc_image.$$"
+        mkdir -p "$MNT_DIR"
+        LOOPBACK="$IMG_TMP_DIR/coal.$$/$$.${OUTPUT_IMG}"
+        OFFSET=$(parted -m "${LOOPBACK}" unit B print | grep fat32 \
+            | cut -f2 -d: | sed 's/.$//')
+        ${SUCMD} mount -o "loop,offset=${OFFSET},uid=${EUID},gid=${GROUPS[0]}" \
+            "${LOOPBACK}" "${MNT_DIR}"
     else
         ${SUCMD} mkdir -p ${MNT_DIR}
         LOOPBACK=$(${SUCMD} lofiadm -a $IMG_TMP_DIR/coal.$$/$$.${OUTPUT_IMG})
         ${SUCMD} mount -F pcfs -o foldcase ${LOOPBACK}:c ${MNT_DIR}
     fi
-    echo "(${IMG_DEV} mounted on ${MNT_DIR}) done."
+    echo "($IMG_TMP_DIR/coal.$$/$$.${OUTPUT_IMG} mounted on ${MNT_DIR}) done."
 }
 
 function copy_config {
@@ -231,6 +242,9 @@ function cleanup
         if [[ -n ${LOOPBACK} ]]; then
             if [[ "$PLATFORM" == "Darwin" ]]; then
                 hdiutil detach ${MNT_DIR} || true
+            elif [[ "$PLATFORM" == "Linux" ]]; then
+                ${SUCMD} umount "${MNT_DIR}" || /usr/bin/true
+                rm -rf "${MNT_DIR}"
             else
                 ${SUCMD} umount ${MNT_DIR} || /usr/bin/true
                 ${SUCMD} lofiadm -d ${LOOPBACK}
diff --git a/bin/build-tar-image b/bin/build-tar-image
index 16b6c7a6..d7e50bc0 100755
--- a/bin/build-tar-image
+++ b/bin/build-tar-image
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright 2015 Joyent, Inc.
+# Copyright 2017 Joyent, Inc.
 #
 
 ROOT=$(cd $(dirname $0)/../; pwd)
@@ -124,7 +124,8 @@ ERROR=0
 CLEANED=0
 
 PLATFORM=$(uname -s)
-if [[ ${PLATFORM} == 'Darwin' || ${PLATFORM} == 'SunOS' ]]; then
+if [[ ${PLATFORM} == 'Darwin' || ${PLATFORM} == 'SunOS' || \
+      ${PLATFORM} == 'Linux' ]]; then
     source ${ROOT}/bin/include-tar-generic
     version
 else
@@ -133,8 +134,8 @@ fi
 
 function test_rootperms
 {
-    # root access is no longer required on OSX
-    [[ ${PLATFORM} == 'Darwin' ]] && return
+    # root access is only required on SunOS
+    [[ ${PLATFORM} != 'SunOS' ]] && return
     su_uid=$(${SUCMD} id -u)
     if [[ ${su_uid} -ne 0 ]]; then
         fatal "Can't get root priviledges."
diff --git a/bin/build-usb-image b/bin/build-usb-image
index 50a450e0..933cf259 100755
--- a/bin/build-usb-image
+++ b/bin/build-usb-image
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright 2015 Joyent, Inc.
+# Copyright 2017 Joyent, Inc.
 #
 
 ROOT=$(cd $(dirname $0)/../; pwd)
@@ -108,8 +108,12 @@ if [[ $($BUILDSPEC -f debug-platform) == 'true' ]]; then
 fi
 TEMPLATE_IMG="$ROOT/cache/file.platimages$PLATFORM_SUFFIX.tgz"
 
-if [[ "$PLATFORM" != "Darwin" ]]; then
+if [[ "$PLATFORM" == "SunOS" ]]; then
     SUCMD='pfexec'
+elif [[ "$PLATFORM" == "Linux" ]]; then
+    SUCMD='sudo'
+    # Parted is required to find the partition offset in the image.
+    [[ $(which parted) ]] || fatal "Please install 'parted'"
 fi
 
 STAGE=${ROOT}/cache/stage_usb
@@ -156,14 +160,20 @@ function mount_image
             | awk '{ print $1 }'`
         MNT_DIR=`grep "Windows_FAT_32" /tmp/output.hdiattach.$$ \
             | awk '{ print $3 }'`
-
+    elif [[ "$PLATFORM" == "Linux" ]]; then
+        MNT_DIR="/tmp/sdc_image.$$"
+        mkdir -p "$MNT_DIR"
+        LOOPBACK=$IMG_TMP_DIR/$$.${OUTPUT_IMG}
+        OFFSET=$(parted -m "${LOOPBACK}" unit B print | grep fat32 \
+            | cut -f2 -d: | sed 's/.$//')
+        ${SUCMD} mount -o "loop,offset=${OFFSET},uid=${EUID},gid=${GROUPS[0]}" \
+            "${LOOPBACK}" "${MNT_DIR}"
     else
         ${SUCMD} mkdir -p ${MNT_DIR}
         LOOPBACK=$(${SUCMD} lofiadm -a $IMG_TMP_DIR/$$.${OUTPUT_IMG})
         ${SUCMD} mount -F pcfs -o foldcase ${LOOPBACK}:c ${MNT_DIR}
     fi
-    # REVIEW - IMG_DEV not defined?
-    echo "(${IMG_DEV} mounted on ${MNT_DIR}) done."
+    echo "($IMG_TMP_DIR/coal.$$/$$.${OUTPUT_IMG} mounted on ${MNT_DIR}) done."
 }
 
 function copy_config {
@@ -196,7 +206,7 @@ function copy_to_mount
             && ${TAR} -c .[a-zA-Z]* * | ${TAR} -C $MNT_DIR -xovf -) \
             || fatal "Unable to copy files to mount"
     else
-        (cd ${MNT_DIR} && ${SUCMD} ${TAR} --no-same-owner \
+        (cd ${MNT_DIR} && ${TAR} --no-same-owner \
          -xvzf $TAR_BUILD_FILENAME) \
             || fatal "Unable to copy files to mount"
     fi
@@ -226,6 +236,9 @@ function cleanup
         if [[ -n ${LOOPBACK} ]]; then
             if [[ "$PLATFORM" == "Darwin" ]]; then
                 hdiutil detach ${MNT_DIR} || true
+            elif [[ "$PLATFORM" == "Linux" ]]; then
+                ${SUCMD} umount "${MNT_DIR}" || /usr/bin/true
+                rm -rf "${MNT_DIR}"
             else
                 ${SUCMD} umount ${MNT_DIR} || /usr/bin/true
                 ${SUCMD} lofiadm -d ${LOOPBACK}
diff --git a/bin/include-tar-generic b/bin/include-tar-generic
index f2580ab2..38bb7318 100644
--- a/bin/include-tar-generic
+++ b/bin/include-tar-generic
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 PLATFORM=$(uname -s)
@@ -25,6 +25,13 @@ case "$PLATFORM" in
     TAR_ROOT=""
     PING="/sbin/ping -c 1"
     ;;
+  Linux)
+    # sudo is not required for tar extraction on Linux
+    SUCMD=''
+    TAR=tar
+    TAR_ROOT=""
+    PING="/bin/ping -c 1"
+    ;;
   SunOS)
     SUCMD='pfexec'
     TAR=gtar
